pkg/virt-controller/watch/pod.go

```diff
@@ -88,21 +88,38 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
 			vmCopy.Status.NodeName = pod.Spec.NodeName
 			// Update the VM
 			logger := logging.DefaultLogger().Object(vm)
-			if err := restClient.Put().Resource("vms").Body(&vmCopy).Name(vmCopy.ObjectMeta.Name).Namespace(kubeapi.NamespaceDefault).Do().Error(); err != nil {
-				logger.Error().Reason(err).Msg("Setting the VM to pending failed.")
-				if e, ok := err.(*errors.StatusError); ok {
-					if e.Status().Reason == metav1.StatusReasonNotFound {
-						// VM does not exist anymore, we don't have to retry
-						return true
-					}
-
-				}
-				logger.V(3).Info().Msg("Enqueuing VM initialization again.")
-				queue.AddRateLimited(key)
-				return true
-			}
+			if putVm(vm, restClient, queue, key) {return true}
+			logger.Info().Msgf("VM successfully scheduled to %s.", vmCopy.Status.NodeName)
+		} else if (vm.Status.Phase == corev1.Running) && (vm.Status.MigrationNodeName == vm.Status.NodeName) {
+			// FIXME: is checking VM status sufficient to identify completed migrations?
+			vmCopy := corev1.VM{}
+			model.Copy(&vmCopy, vm)
+			vmCopy.Status.MigrationNodeName = nil
+			logger := logging.DefaultLogger().Object(vm)
+			// FIXME: should the associated migration be deleted here?
+			if putVm(vmCopy, restClient, queue, key) {return true}
 			logger.Info().Msgf("VM successfully scheduled to %s.", vmCopy.Status.NodeName)
 		}
 		return true
 	})
 }
+
+// syncronously post updated VM object to API server.
+// requeue on error if appropriate
+func putVm(vm corev1.VM, restClient *rest.RESTClient, queue workqueue.RateLimitingInterface, key interface{}) error {
+	logger := logging.DefaultLogger().Object(vm)
+	if err := restClient.Put().Resource("vms").Body(&vm).Name(vm.ObjectMeta.Name).Namespace(kubeapi.NamespaceDefault).Do().Error(); err != nil {
+		logger.Error().Reason(err).Msg("Setting the VM state failed.")
+		if e, ok := err.(*errors.StatusError); ok {
+			if e.Status().Reason == metav1.StatusReasonNotFound {
+				// VM does not exist anymore, we don't have to retry
+				return nil
+			}
+
+		}
+		logger.V(3).Info().Msg("Enqueuing VM again.")
+		queue.AddRateLimited(key)
```

Comment:
We do. But I would prefer if you return the error and do the queue.AddRateLimited in the main controller loop.